### ISSUE TITLE
Multioutput regression with custom loss function (least squares)

### ISSUE THREAD

**Issue by SimonCalo** [dmlc/xgboost#5488]
Hello everyone!

I am very new to ML and even more so to XGBoost, so please excuse me if I ignore some obvious piece of knowledge.

I am currently working on a problem where I am trying to predict multiple outputs. To do so, I am going to try to embed the XGBoost regression within the SciKit learn multioutput class. (I still haven't tried this because I wanted to first structure the procedure nicely).

However, I would like the predictions to minimize the error on a quantity obtained by a combination of the predictions. To clarify, this quantity is the 3-momentum magnitude given the 3 spacial components. To estimate the error during the training, I would like to rely on simple least-squares but on the momentum magnitude rather than the 3 individual components.

I saw that this could be achieved by creating a custom loss function that uses the outputs and calculates the gradient and the hessian in the following way:
``` python
def gradient(y_pred, y_true):
  return 2*(y_pred-y_true)

def hessian(y_pred, y_true):
  return 0*y_pred + 2

def custom_objective (predict, truth):

  truth_px = truth[:, 0]
  truth_py = truth[:, 1]
  truth_pz = truth[:, 2]
  truth_p = math.sqrt((truth_px)**2 + (truth_py)**2 +(truth_pz)**2)
  
  predict_px = predict[:, 0]
  predict_py = predict[:, 0]
  predict_pz = predict[:, 0]
  predict_p = math.sqrt((predict_px)**2 + (predict_py)**2 + (predict_pz)**2)
  
  grad = gradient(predict_p, truth_p)
  hess = hessian(predict_p, truth_p)
  return grad, hess
```
However, I am still unsure whether this would be the correct implementation to achieve my goal or if there was a more straightforward way.

Thank you all in advance for your suggestions!

**Comment by SimonCalo** [comment]
As a follow-up, I have been trying to make it work, but there seems to be a major issue:
The SciKit learn wrapper class MultiOutputRegressor essentially fits one regressor at a time, as stated also in the class documentation: "This strategy consists of fitting one regressor per target".

Now, this means that in my custom loss function, I can only access one momentum direction at a time, instead of accessing them all at once. 

At the moment, I am very unsure how to get around this problem and any help or suggestion would be truly appreciated.

For reference, I am pasting my whole code here:

`import root_pandas
import xgboost as xgb
from xgboost import XGBRegressor
from sklearn.multioutput import MultiOutputRegressor
import matplotlib.pyplot as plt
import numpy as np
from custom_obj import *

data_file = root_pandas.read_root("../diamond_new.root")

subset = data_file[['Dst_px', 'Dst_py','Dst_pz', 'Dst_theta', 'l_theta', 'l_px', 'l_py','l_pz', 'B0_DiamondFrame_px', 'B0_DiamondFrame_py', 'B0_DiamondFrame_pz', '_DiamondFrame_q2', '_DiamondFrame_w', 'Dst_phi', 'l_phi']]

train_x = subset[:65000].values
train_y = data_file[['B0_px', 'B0_py', 'B0_pz']][:65000].values

clf = MultiOutputRegressor(XGBRegressor(objective=custom_objective, silent=False, learning_rate = 0.01, n_estimators=1000, max_depth=4))


clf.fit(train_x, train_y)`

With custom_obj being a module containing the following functions:

`import math

def gradient(y_pred, y_true):
  return 2*(y_pred-y_true)

def hessian(y_pred, y_true):
  return 0*y_pred + 2

def custom_objective (train_y, y_pred):

  truth_px = train_y[:,0]
  truth_py = train_y[:,1]
  truth_pz = train_y[:,2]
  truth_p = math.sqrt((truth_px)**2 + (truth_py)**2 +(truth_pz)**2)

  predict_px = y_pred[:,0]
  predict_py = y_pred[:,1]
  predict_pz = y_pred[:,2]
  predict_p = math.sqrt((predict_px)**2 + (predict_py)**2 + (predict_pz)**2)

  grad = gradient(predict_p, truth_p)
  hess = hessian(predict_p, truth_p)
  return grad, hess`

**Comment by hcho3** [comment]
@SimonCalo Unfortunately, it is currently not possible to perform regression with multiple outputs using XGBoost. We are currently working on a proof-of-concept implementation of multi-output regression: #5460. In the meanwhile, you should consider using deep learning frameworks such as PyTorch.

**Comment by trivialfis** [comment]
@hcho3  It's possible with 1 stack of trees per target (which is the case in this issue).  I haven't had the energy to go through what's in this issue.  Can we reopen it?

**Comment by hcho3** [comment]
@trivialfis I am re-opening this issue per your request.

The OP (@SimonCarlo) said that fitting one stack per target does not serve his needs. 

Quote:

> The SciKit learn wrapper class MultiOutputRegressor essentially fits one regressor at a time, as stated also in the class documentation: "This strategy consists of fitting one regressor per target".
Now, this means that in my custom loss function, I can only access one momentum direction at a time, instead of accessing them all at once.

**Comment by reseal** [comment]
@hcho3 @trivialfis what is the latest on the possibility of doing this? Multi-output regression is supported since v1.6. To use a custom objective, will it be as simple as having the custom objective function output 2d and 3d-tensors for the gradient and hessian, respectively?

**Comment by AhmedThahir** [comment]
Any updates on this?

**Comment by trivialfis** [comment]
For model per target, yes, it's supported,  for both CPU (numpy) gradient and GPU (cupy)

**Comment by AhmedThahir** [comment]
Is it possible to perform maximum likelihood estimation?

I want the model to output mu and sigma, and minimize the negative log likelihood using a custom loss.

**Comment by trivialfis** [comment]
Yes, you can derive the gradient for forecasting objectives.

**Comment by AhmedThahir** [comment]
Could you share an example, or a guide which I can refer to? I've not been able to figure it out.

**Comment by trivialfis** [comment]
I can give some examples but it will take time to dig up some old notes, need to sort out other things first.

In general, it's just taking partial derivatives on the MLE for gradient and use Fisher info in place of Hessian.  The trick is to reparametrize using exp to make sure variance is greater than 0.

**Comment by trivialfis** [comment]
An in-depth example can be found in the master branch. https://github.com/dmlc/xgboost/blob/master/doc/tutorials/advanced_custom_obj.rst

**Comment by AhmedThahir** [comment]
> An in-depth example can be found in the master branch. https://github.com/dmlc/xgboost/blob/master/doc/tutorials/advanced_custom_obj.rst

Thanks, but the example shown seems to be a single output? What if I want to output the mean and variance for a given x?

**Comment by AhmedThahir** [comment]
> I can give some examples but it will take time to dig up some old notes, need to sort out other things first.

No issues, please take your time.

> In general, it's just taking partial derivatives on the MLE for gradient and use Fisher info in place of Hessian. The trick is to reparametrize using exp to make sure variance is greater than 0.

I am not able to follow this, a code snippet wrt multi-output would be better
